warehouse/retry.go

```diff
@@ -0,0 +1,125 @@
+package warehouse
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"github.com/rudderlabs/rudder-server/proto/warehouse"
+	"strings"
+)
+
+type RetryReq struct {
+	WorkspaceID   string
+	DestinationID string
+
+	IntervalInHours int64   // Optional, if provided we will retry based on the interval provided
+	UploadIds       []int64 // Optional, if provided we will retry the upload ids provided
+
+	API        UploadAPIT
+	ForceRetry bool
+}
+
+func (retryReq *RetryReq) RetryWHUploads() (response *proto.RetryWHUploadsResponse, err error) {
+	// Request validation
+	err = retryReq.validateReq()
+	defer func() {
+		if err != nil {
+			retryReq.API.log.Errorf("WH: Error occurred while retrying upload jobs with error: %s", err.Error())
+			response = &proto.RetryWHUploadsResponse{
+				Message:    err.Error(),
+				StatusCode: 400,
+			}
+		}
+	}()
+	if err != nil {
+		return
+	}
+
+	// Generating query
+	retryQuery := retryReq.queryToRetry()
+
+	// Retrying uploads
+	uploadsRetried, err := retryReq.retryUploads(retryQuery)
+	if err != nil {
+		return
+	}
+
+	response = &proto.RetryWHUploadsResponse{
+		Message:    retryReq.successMessage(uploadsRetried),
+		StatusCode: 200,
+	}
+	return
+}
+
+func (retryReq *RetryReq) queryToRetry() (sqlStatement string) {
+	var retryClause, statusClause string
+	if len(retryReq.UploadIds) != 0 {
+		retryClause = fmt.Sprintf(`id IN (%s)`, strings.Trim(strings.Replace(fmt.Sprint(retryReq.UploadIds), " ", ",", -1), "[]"))
+	} else {
+		retryClause = fmt.Sprintf(`destination_id = '%s' AND created_at > NOW() - INTERVAL '%d HOUR'`, retryReq.DestinationID, retryReq.IntervalInHours)
+	}
+	if !retryReq.ForceRetry {
+		statusClause = fmt.Sprintf("AND status = '%s'", Aborted)
+	}
+
+	sqlStatement = fmt.Sprintf(`
+		UPDATE wh_uploads
+		SET
+			metadata = metadata || '{"retried": true, "priority": 50}' || jsonb_build_object('nextRetryTime', NOW() - INTERVAL '1 HOUR'),
+			status = 'waiting',
+			updated_at = NOW()
+		WHERE %[1]s %[2]s
+        RETURNING id`,
+		retryClause,
+		statusClause,
+	)
+	retryReq.API.log.Info(sqlStatement)
```

Comment:
As discussed, let's move it to `golang` sql package's view of handling parameters to prevent sql injection ?

Reply:
Yes.
Will try this today.